### DISCUSSION TITLE
Different range strategies based on python require type

### DISCUSSION THREAD

**Discussion by kasium** [renovatebot/renovate#30715] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gerrit; v38

### Please tell us more about your question or problem

Given the following setup.py

```py
if __name__ == "__main__":
  setup(
    install_requires = ["lib1~=1.1"],
    extras_require = {
      "test": ["libB==1.2.3"]
    }
```

I would like to achive the following:

- for all dependencies in install_requires use the `widen` range strategy
- for all others use the `replace` range strategy

Of course I could define rules for dependencies with install_requires but its a burdon to keep the list up to date.
Any better idea

**Comment by RahulGautamSingh** [comment, marked as answer]
Maybe you could use `matchDepTypes` here, if `depType` of the deps in `install_requires` and `extras_requires` are different.

You can find the `depType` of the packages in the logs. 
I don't think there's another way except that if the `depTypes` turn out to be the same or the value is not available.

**Reply by kasium** [reply]
Cool! I don't see any way to see the extra key name, so in the above example test

**Reply by RahulGautamSingh** [reply]
Check for it in the logs, when you run the bot on you repo.

**Reply by kasium** [reply]
Right, I see. However, for pep621, it does not show the context but only

```
.639711237Z                {
2024-08-12T10:22:27.639713933Z                  "packageName": "invoke",
2024-08-12T10:22:27.639716872Z                  "depName": "invoke",
2024-08-12T10:22:27.639719606Z                  "datasource": "pypi",
2024-08-12T10:22:27.639722337Z                  "depType": "project.optional-dependencies",
2024-08-12T10:22:27.639725052Z                  "currentValue": "==2.2.0",
2024-08-12T10:22:27.639727753Z                  "currentVersion": "2.2.0",
2024-08-12T10:22:27.639730411Z                  "updates": [],
2024-08-12T10:22:27.639733073Z                  "versioning": "pep440",
2024-08-12T10:22:27.639735758Z                  "warnings": [],
2024-08-12T10:22:27.639738660Z                  "fixedVersion": "2.2.0"
2024-08-12T10:22:27.639741448Z                },
```

**Reply by RahulGautamSingh** [reply]
Attach your full logs, please.

**Reply by kasium** [reply]
[logs.txt](https://github.com/user-attachments/files/16580943/logs.txt)

**Reply by RahulGautamSingh** [reply]
You mentioned `setup.py` in the description but the logs are for `pyproject.toml`, which one is it?

Update: You have `enabledManagers: ['pep621']` in your config. Change that to `enabledManagers: ['pep621', 'pip_setup']` so that `setup.py` file deps are also extracted.

**Reply by kasium** [reply]
Right, but as written above `However, for pep621, it does not show the context but only` -> the logs are for this case

Meaning, for pep621 I can only see if it comes from the install or extra dependencies but not from which one

**Reply by RahulGautamSingh** [reply]
sorry, what do you mean by context? Which other info do you want to see?

**Reply by kasium** [reply]
So, the optional dependencies in the pyproject.toml look like this
```toml
[project.optional-dependencies]
test= ["pytest~=1.10"]
```

Currently, renovate only shows `project.optional-dependencies`, but it would be awesome to show `project.optional-dependencies.test`

**Reply by RahulGautamSingh** [reply]
Okay got it. Please create a new discussion of category Suggest an Idea with the logs and your use case.
